File: src/nuitrack_bridge/types/skeleton_data.cc
#include "nuitrack_bridge/types/skeleton_data.h"
#include "nuitrack/types/SkeletonData.h"

namespace nuitrack_bridge::skeleton_data {

    int32_t getNumSkeletons(const SkeletonData& frame) {
        return static_cast<int32_t>(frame.getNumSkeletons());
    }

    std::unique_ptr<std::vector<Skeleton>> getSkeletons(const SkeletonData& frame) {
        return std::make_unique<std::vector<Skeleton>>(frame.getSkeletons());
    }

    uint64_t getTimestamp(const SkeletonData& frame) {
        return frame.getTimestamp();
    }

}

File: src/nuitrack_bridge/types/skeleton_data.rs
#[cxx::bridge(namespace = "nuitrack_bridge::skeleton_data")]
pub mod ffi {

    #[namespace = "tdv::nuitrack"]
    unsafe extern "C++" {
        type SkeletonData;
        type Skeleton = crate::nuitrack_bridge::types::skeleton::ffi::Skeleton;
    }

    impl SharedPtr<SkeletonData> {}

    unsafe extern "C++" {
        include!("nuitrack_bridge/types/skeleton_data.h");
        
        #[cxx_name = "getNumSkeletons"]
        pub fn num_skeletons(skeleton_frame: &SkeletonData) -> Result<i32>;

        #[cxx_name = "getSkeletons"]
        pub fn skeletons(skeleton_frame: &SkeletonData) -> Result<UniquePtr<CxxVector<Skeleton>>>;

        #[cxx_name = "getTimestamp"]
        pub fn timestamp(skeleton_frame: &SkeletonData) -> Result<u64>;

    }
}

unsafe impl Send for ffi::SkeletonData {}
unsafe impl Sync for ffi::SkeletonData {}